src/rust/src/pkcs7.rs

```diff
@@ -162,6 +166,116 @@ fn encrypt_and_serialize<'p>(
     }
 }
 
+#[pyo3::pyfunction]
+fn deserialize_and_decrypt<'p>(
+    py: pyo3::Python<'p>,
+    data: CffiBuf<'p>,
+    certificate: pyo3::Bound<'p, x509::certificate::Certificate>,
+    private_key: pyo3::Bound<'p, backend::rsa::RsaPrivateKey>,
+    encoding: &pyo3::Bound<'p, pyo3::PyAny>,
+    options: &pyo3::Bound<'p, pyo3::types::PyList>,
+) -> CryptographyResult<pyo3::Bound<'p, pyo3::types::PyBytes>> {
+    // Extract the data on the right format (PEM or DER, SMIME handled by Python)
+    let extracted_data = if encoding.is(&types::ENCODING_DER.get(py)?) {
+        data.as_bytes().to_vec()
+    } else {
+        let pem_str = std::str::from_utf8(data.as_bytes())
+            .map_err(|_| pyo3::exceptions::PyValueError::new_err("Invalid PEM data"))?;
+        let pem = pem::parse(pem_str)
+            .map_err(|_| pyo3::exceptions::PyValueError::new_err("Failed to parse PEM data"))?;
+        pem.contents().to_vec()
+    };
+
+    // Deserialize the content info
+    let content_info = asn1::parse_single::<pkcs7::ContentInfo<'_>>(&extracted_data).unwrap();
```

Comment:
@reaperhulk This would be our first time using our DER parser for PKCS7. On the one hand: new API, so there's no backwards-compat concerns. On the other hand, _surely_ someone is emitting awful BER PKCS#7. Any concerns?

Reply:
@reaperhulk just going to keep pinging you on this :-) (Enjoy your vacation)

Reply:
Let's assume we release with DER only parsing. In that scenario we're attempting to empirically determine whether or not there's enough BER out there to matter. If no one comes to complain then great, job done. However, if we discover we're wrong and BER is relevant, what is our plan? Are we going to add BER support in rust?

Reply:
If we do determine that this API is basically low utility for many users because their S/MIME messages aren't valid DER, it seems we have a few options:

1) Declare that's how it goes. This is a net new API, so no users are worse off than they were before. May result in greater timelines for adoption of this, and migration away from other libraries.

2) Re-implement in terms of OpenSSL S/MIME APIs. Probably straightforward, although as-ever, assessing the compatibility surface will be a PITA.

3) Add BER support to rust-asn1. No fucking chance, not worth discussing.

4) We force ourselves to do a survey of the ecosystem before landing this so we can contact producers of non-DER S/MIMEs.


I have a soft spot for 1, I could be persuaded of 4, and if we really need to do it, I guess 2 is the way to go.

WDYT?

Reply:
Go stdlib doesn't support PKCS#7 I don't think. I did find:

- https://github.com/mozilla-services/pkcs7/blob/v0.9.0/pkcs7.go#L155-L160
- https://github.com/InfiniteLoopSpace/go_S-MIME/blob/master/cms/protocol/contentinfo.go#L25


Sooooo

Reply:
Okay, I'm comfortable with doing this DER only and finding out how it goes. It's net-new API surface so we're not regressing and I'd love to find out BER doesn't matter.